### PR TITLE
Minimal working charm

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: https://docs.google.com/document/d/13K6TlXI2b_Hg0-0_b0LeMhPJ3WVdfJtFPsImVRgiFNc

### Overview

Partial implementation of the spect above, providing a minimal working SMTP charm without secrets support

### Rationale

N/A

### Juju Events Changes

* Added relation created
* Added config changed
* Added update status

### Module Changes

Added charm.py and charm_state.py modules

### Library Changes

New smtp library

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

Charmhub docs will be provided when the spec above is fully implemented

### CODE REVIEW

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+This library contains the Requires and Provides classes for handling the relation
+between an application and a charm providing the `smtp`relation.
+It also contains a `SmtpRelationData` class to wrap the SMTP data that will
+be shared via the relation.
+
+### Requirer Charm
+
+```python
+
+from charms.mtp_integrator.v0 import SmtpDataAvailableEvent, SmtpRequires
```
Comment (amandahla) [inline lib/charms/smtp_integrator/v0/smtp.py:17]: charms.smtp_integrator

File: metadata.yaml [hunk 2]
```diff
@@ -1,50 +1,31 @@
...
-# (Recommended)
-display-name: Charm Template
-
-# (Required)
-summary: A very short one-line summary of the charm.
+name: smtp-integrator
+display-name: SMTP Integrator
+summary: SMTP Integrator to centralize SMTP configuration.
 docs: https://discourse.charmhub.io
 issues: https://github.com/canonical/is-charms-template-repo/issues
```
Comment (amandahla) [inline metadata.yaml:7]: s/is-charms-template-repo/smtp-integrator-operator

File: requirements.txt [hunk 3]
```diff
@@ -1 +1,2 @@
-ops >= 2.2.0
+ops==2.6.0
```
Comment (amandahla) [inline requirements.txt:1]: Could it be ops 2.7.0 ?

File: src/charm.py [hunk 4]
```diff
@@ -33,83 +32,59 @@ def __init__(self, *args):
...
+            self.smtp.update_relation_data(relation, self.get_smtp_data())
 
-        Learn more about config at https://juju.is/docs/sdk/config
+    def get_smtp_data(self) -> smtp.SmtpRelationData:
+        """Get relation data.
 
-        Args:
-            event: event triggering the handler.
+        Returns:
+            SmtpRelationData containing the IdP details.
```
Comment (amandahla) [inline src/charm.py:76]: nitpick: IdP details?

File: src/charm_state.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class CharmState:
```
Comment (amandahla) [inline src/charm_state.py:64]: Recently I had a review on Synapse to make CharmState a dataclass. What do you think?

**Review by amandahla — COMMENTED** [review]
Nice 🥳 Left a few comments.

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        self.smtp = smtp.SmtpRequires(self)
+        self.framework.observe(self.smtp.on.smtp_data_available, self._handler)
+        ...
+
+    def _handler(self, events: SmtpDataAvailableEvent) -> None:
+        ...
+
+```
+
+As shown above, the library provides a custom event to handle the sceneario in
```
Comment (varshigupta12) [inline lib/charms/smtp_integrator/v0/smtp.py:31]: Nit: scenario*

File: metadata.yaml [hunk 7]
```diff
@@ -1,50 +1,31 @@
...
   - https://launchpad.net/~canonical-is-devops
-source: https://github.com/canonical/is-charms-template-repo
-
+source: https://github.com/canonical/smtp-integrator-operator/issues
 description: |
-  A single sentence that says what the charm is, concisely and memorably.
-
-  A paragraph of one to three short sentences, that describe what the charm does.
+  A [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators)
+  deploying and managing a SMTP Integrator on Kubernetes. SMTP s a communication
```
Comment (varshigupta12) [inline metadata.yaml:13]: SMTP is*

File: metadata.yaml [hunk 8]
```diff
@@ -1,50 +1,31 @@
...
-name: is-charms-template
- 
-# The following metadata are human-readable and will be published prominently on Charmhub.
-
-# (Recommended)
-display-name: Charm Template
-
-# (Required)
-summary: A very short one-line summary of the charm.
+name: smtp-integrator
```
Comment (varshigupta12) [inline metadata.yaml:3]: Is the intention for this charm to be k8s only? In that case, we should add a suffix '-k8s' at the end of the name.

File: requirements.txt (hunk 3 above)
Comment (arturo-seijas) [reply requirements.txt:1]: It could, yes ;)

File: src/charm_state.py (hunk 5 above)
Comment (arturo-seijas) [reply src/charm_state.py:64]: Makes sense

File: metadata.yaml (hunk 8 above)
Comment (arturo-seijas) [reply metadata.yaml:3]: No, it can be deployed in both substrates

**Review by amandahla — DISMISSED** [review]
LGTM

File: lib/charms/smtp_integrator/v0/smtp.py
Comment (cbartz) [inline lib/charms/smtp_integrator/v0/smtp.py:1]: I would prefer it if the docstring (which will be used as documentation for charmhub, I suppose) could use the term `integration` instead of `relation`.

**Review by cbartz — COMMENTED** [review]
Looks good! I have a couple of minor comments. 
I also noticed that the term `relation` is still used a lot for variables and function names. Maybe the term `integration` would be better, as the term will be outdated in the future, but this might be something to agree with the team.

File: src/charm_state.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    "port",
+    "user",
+    "password",
+    "auth_type",
+    "transport_security",
+    "domain",
+)
+
+
+class SmtpIntegratorConfig(BaseModel):  # pylint: disable=too-few-public-methods
```
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply src/charm_state.py:27]: Good point

File: lib/charms/smtp_integrator/v0/smtp.py
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:1]: Makes sense

Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:1]: Considering the RelationChanged and the RelationEvent are still named after the old terminology, I've left some of the occurrences of the term to avoid confusion.

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+        host: The hostname or IP address of the outgoing SMTP relay.
+        port: The port of the outgoing SMTP relay.
+        user: The SMTP AUTH user to use for the outgoing SMTP relay.
+        password: The SMTP AUTH password to use for the outgoing SMTP relay.
+        auth_type: The type used to authenticate with the SMTP relay.
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    host: str = Field(..., min_length=1)
```
Comment (gregory-schiano) [inline lib/charms/smtp_integrator/v0/smtp.py:114]: I'm wondering if the validation couldn't be a bit more specific, I know pydantic supports IPv4 and IPv6 address validation, we should look for "hostname" validation and use a Union

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    port: int
+    user: Optional[str]
+    password: Optional[str]
+    auth_type: AuthType
+    transport_security: TransportSecurity
+    domain: Optional[str]
```
Comment (gregory-schiano) [inline lib/charms/smtp_integrator/v0/smtp.py:120]: If there's a hostname validation, we could use it here too

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        assert event.relation.app
+        if event.relation.data[event.relation.app]:
+            self.on.smtp_data_available.emit(event.relation, app=event.app, unit=event.unit)
```
Comment (gregory-schiano) [inline lib/charms/smtp_integrator/v0/smtp.py:240]: I think the data should be validated here too

File: metadata.yaml [hunk 13]
```diff
@@ -1,50 +1,31 @@
...
 description: |
-  A single sentence that says what the charm is, concisely and memorably.
-
-  A paragraph of one to three short sentences, that describe what the charm does.
+  A [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators)
+  deploying and managing a SMTP Integrator on Kubernetes and bare metal. SMTP is a communication
+  protocol used for sending and receiving email messages over the Internet.
 
-  A third paragraph that explains what need the charm meets.
+  This charm simplifies configuration of SMTP by providing a single point
```
Comment (gregory-schiano) [inline metadata.yaml:16]: Could be worthwhile to specify it's a workload-less charm

File: metadata.yaml [hunk 14]
```diff
@@ -1,50 +1,31 @@
...
   - https://launchpad.net/~canonical-is-devops
-source: https://github.com/canonical/is-charms-template-repo
-
+source: https://github.com/canonical/smtp-integrator-operator/issues
 description: |
-  A single sentence that says what the charm is, concisely and memorably.
-
-  A paragraph of one to three short sentences, that describe what the charm does.
+  A [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators)
+  deploying and managing a SMTP Integrator on Kubernetes and bare metal. SMTP is a communication
```
Comment (gregory-schiano) [inline metadata.yaml:13]: More a question, is it relevant to run this charm as a machine charm ?

File: metadata.yaml [hunk 15]
```diff
@@ -1,50 +1,31 @@
...
-# Remove them if not required.
+  For DevOps or SRE teams this charm will make operating any charm leveraging the SMTP
+  protocol simple and straightforward through Juju's clean interface.
 
-# Your workload’s containers.
-containers:
-  httpbin:
-    resource: httpbin-image
+provides:
+  smtp-legacy:
```
Comment (gregory-schiano) [inline metadata.yaml:29]: Why is it called `smtp-legacy` ?

File: src/charm.py [hunk 16]
```diff
@@ -33,83 +32,59 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
```
Comment (gregory-schiano) [inline src/charm.py:39]: I'm wondering if `CharmState` is relevant here as it's a copy of `SmtpRelationData` validation, so doesn't add any value

File: tests/integration/test_charm.py
Comment (gregory-schiano) [inline tests/integration/test_charm.py:1]: We should probably also add a test where we set and invalid configuration value, and ensure the charm is in blocked state with a message

Comment (gregory-schiano) [reply tests/integration/test_charm.py:1]: And also probably with an empty config

File: tests/unit/test_charm_state.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+from unittest.mock import MagicMock
+
+import pytest
+import yaml
+
+from charm_state import KNOWN_CHARM_CONFIG, CharmConfigInvalidError, CharmState
+
+
+def test_known_charm_config():
```
Comment (gregory-schiano) [inline tests/unit/test_charm_state.py:14]: Nice I like this test 👍

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+        port: The port of the outgoing SMTP relay.
+        user: The SMTP AUTH user to use for the outgoing SMTP relay.
+        password: The SMTP AUTH password to use for the outgoing SMTP relay.
+        auth_type: The type used to authenticate with the SMTP relay.
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    host: str = Field(..., min_length=1)
+    port: int
```
Comment (gregory-schiano) [inline lib/charms/smtp_integrator/v0/smtp.py:115]: More validation can be set on the port (> 0 and <65536)

File: tests/unit/test_library_smtp.py
Comment (gregory-schiano) [inline tests/unit/test_library_smtp.py:1]: I think we should also test here all the validation rules with edge cases and non happy cases:
- IPv4 address
- IPv6 address
- Invalid port (<0 or >65535)
- Invalid enum values
...

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (gregory-schiano) [inline lib/charms/smtp_integrator/v0/smtp.py:240]: I think also we should probably write a warning log when insecure settings are set like "auth_type == none" or "tls_security == none"
We allow it, but warn the user these are unsafe settings and should be use with caution

File: metadata.yaml (hunk 14 above)
Comment (arturo-seijas) [reply metadata.yaml:13]: It might be for the operator if deploying everything on VMs/bare metal

File: metadata.yaml (hunk 15 above)
Comment (arturo-seijas) [reply metadata.yaml:29]: That has been defined in the spec. `smtp` will support secrets and share a secret id instead of the password

File: tests/integration/test_charm.py
Comment (arturo-seijas) [reply tests/integration/test_charm.py:1]: That's already unit tested

File: metadata.yaml (hunk 13 above)
Comment (arturo-seijas) [reply metadata.yaml:16]: Perhaps that's too detailed for the metadata? I'd rather specify that in the documentation, when delivered, as part of the `architecture` docs

File: src/charm.py (hunk 16 above)
Comment (arturo-seijas) [reply src/charm.py:39]: I expect it to diverge when the `smtp` relation is added, so perhaps we can defer the decision until then

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 18 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:115]: Done

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:249]: Added

Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py]: You mean the relation content? It's already validated before converting it to strings

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 10 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:117]: I haven't found any "hostname" validation, but perhaps we could use a regex

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (gregory-schiano) [reply lib/charms/smtp_integrator/v0/smtp.py]: But it's validated on the Provider side, not the required. And we have no guarantee that the Provider library was used when consuming the data from the Requirer. So as a rule of thumbs, always validate the data from both sides

File: metadata.yaml (hunk 13 above)
Comment (gregory-schiano) [reply metadata.yaml:16]: Ok

File: metadata.yaml (hunk 14 above)
Comment (gregory-schiano) [reply metadata.yaml:13]: Then we should probably rephrase the next sentences which are very focused on K8S

File: metadata.yaml (hunk 15 above)
Comment (gregory-schiano) [reply metadata.yaml:29]: Forgot about this, fine

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py]: Perhaps we should instantiate a `SmtpRelationData` object here and merge this object with `SmtpDataAvailableEvent`. Otherwise we'd be duplicating the code

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 10 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:117]: I'm not sure if it's worth, tbh

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+import logging
+from enum import Enum
+from typing import Optional
+
+import ops
+from pydantic import BaseModel, Field
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "smtp-legacy"
```
Comment (jdkandersson) [inline lib/charms/smtp_integrator/v0/smtp.py:73]: Out of curiosity, what is the reason for calling this legacy please?

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+    def transport_security(self) -> str:
+        """Fetch the SMTP transport security protocol from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("transport_security")
+
+    @property
+    def domain(self) -> str:
+        """Fetch the SMTP domain from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("domain")
```
Comment (jdkandersson) [inline lib/charms/smtp_integrator/v0/smtp.py:199]: @gregory-schiano and @benhoyt as well for thoughts, these properties have the possibility of raising `AssertionError` (not sure under what conditions `self.relation.app` could be non-truthy). Additionally, the `.get` functions could return `None` which the type hints doesn't seem to indicate. Any thoughts on that?

Comment (jdkandersson) [reply lib/charms/smtp_integrator/v0/smtp.py:199]: Perhaps the easiest way to solve this would be to make these functions and raise some sort of custom error if `self.relation.app` is falsy or when the `.get` returns `None`

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        assert event.relation.app
```
Comment (jdkandersson) [inline lib/charms/smtp_integrator/v0/smtp.py:241]: Here, and anywhere else that has an `assert`, it is generally more user friendly to raise an exception rather than a generic `AssertionError` since the custom exception can be specifically handled whereas a general `AssertionError` must be generically handled

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+        relation_data = event.relation.data[event.relation.app]
+        if relation_data:
+            print(relation_data)
```
Comment (jdkandersson) [inline lib/charms/smtp_integrator/v0/smtp.py:244]: Is this `print` still required?

File: src-docs/charm_state.py.md [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+
+
+
+---
+
+## <kbd>class</kbd> `CharmState`
+Represents the state of the SMTP Integrator charm. 
+
+Attrs:  host: The hostname or IP address of the outgoing SMTP relay.  port: The port of the outgoing SMTP relay.  user: The SMTP AUTH user to use for the outgoing SMTP relay.  password: The SMTP AUTH password to use for the outgoing SMTP relay.  auth_type: The type used to authenticate with the SMTP relay.  transport_security: The security protocol to use for the outgoing SMTP relay.  domain: The domain used by the sent emails from SMTP relay.
```
Comment (jdkandersson) [inline src-docs/charm_state.py.md:45]: This needs to be called `Attributes` so that it renders properly here please

File: src/charm.py [hunk 24]
```diff
@@ -33,83 +31,59 @@ def __init__(self, *args):
...
-        this method.
+    def _update_relations(self) -> None:
+        """Update all SMTP data for the existing relations."""
+        if not self.model.unit.is_leader():
+            return
+        for relation in self.smtp.relations:
+            self.smtp.update_relation_data(relation, self.get_smtp_data())
 
-        Learn more about config at https://juju.is/docs/sdk/config
+    def get_smtp_data(self) -> smtp.SmtpRelationData:
```
Comment (jdkandersson) [inline src/charm.py:71]: What is the reason for this being a public method please? Could it be private?

File: src/charm_state.py [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+        Raises:
+            CharmConfigInvalidError: if the charm configuration is invalid.
+        """
+        config = {k: v for k, v in charm.config.items() if k in KNOWN_CHARM_CONFIG}
+        try:
+            # Incompatible with pydantic.AnyHttpUrl
+            valid_config = SmtpIntegratorConfig(**config)  # type: ignore
+        except ValidationError as exc:
+            error_fields = set(
+                itertools.chain.from_iterable(error["loc"] for error in exc.errors())
```
Comment (jdkandersson) [inline src/charm_state.py:120]: Out of curiosity, what would be an example value of `error["loc"]` in this case please?

**Review by jdkandersson — COMMENTED** [review]
Up to the tests

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 20 above)
Comment (benhoyt) [reply lib/charms/smtp_integrator/v0/smtp.py:199]: It looks like `Relation.app` should "basically never" be None. I know that's not very convincing :-) ... but the code tries very hard to set that. I've opened an [issue](https://github.com/canonical/operator/issues/1040) to investigate further and see if we can change the type annotation for `Relation.app` to be straight `Application` (that is, non-optional).

As the type annotation stands, however, I think the `assert` is exactly right: we're telling the type checker (and Python) we are sure this shouldn't be None.

A helper function `relation_app_data` that does this sounds reasonable though.

Comment (jdkandersson) [reply lib/charms/smtp_integrator/v0/smtp.py:199]: If it is just aimed at type hints then [`typing.cast`](https://docs.python.org/3/library/typing.html#typing.cast) might be better because it can't raise an assertion error

It would be good if the type could be changed to be non optional if it can basically never be `None` 🙂

Comment (benhoyt) [reply lib/charms/smtp_integrator/v0/smtp.py:199]: > It would be good if the type could be changed to be non optional if it can basically never be None 🙂

Agreed. We'll aim to do that.

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 19 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:73]: You can find the details in the spec

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 22 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py]: Removed

File: src-docs/charm_state.py.md (hunk 23 above)
Comment (arturo-seijas) [reply src-docs/charm_state.py.md]: Noted. Thanks

File: src/charm_state.py (hunk 25 above)
Comment (arturo-seijas) [reply src/charm_state.py:120]: `loc` is, per the pydantic docs, the error location, that is, the key whose value does not pass validation, i.e. `host`

File: src/charm.py (hunk 24 above)
Comment (arturo-seijas) [reply src/charm.py]: Indeed. Changed

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 10 above)
Comment (gregory-schiano) [reply lib/charms/smtp_integrator/v0/smtp.py:117]: Let's keep it like this, regex could be a mess in some cases

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (gregory-schiano) [reply lib/charms/smtp_integrator/v0/smtp.py]: That's fine, indeed better to avoid duplicating the validation

File: src/charm.py (hunk 16 above)
Comment (gregory-schiano) [reply src/charm.py:38]: Fine for me

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 20 above)
Comment (gregory-schiano) [reply lib/charms/smtp_integrator/v0/smtp.py:199]: This assertion doesn't bother me too much, raising an assertion error is not ideal but allows us to catch cases that shouldn't happen which is the goal of asserts in such kind of code.
`typing.cast` is also a valid approach

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 11 above)
Comment (gregory-schiano) [reply lib/charms/smtp_integrator/v0/smtp.py:123]: As discussed previously let's put hostname validation aside

File: config.yaml [hunk 26]
```diff
@@ -1,16 +1,28 @@
...
+  user:
+    description: The SMTP AUTH user to use for the outgoing SMTP relay
+    type: string
+  password:
+    description: The SMTP AUTH password to use for the outgoing SMTP relay
+    type: string
+  auth_type:
+    description: The type used to authenticate with the SMTP relay. It has to be one of "none", "not_provided" or "plain"
+    type: string
+    default: "none"
```
Comment (yanksyoon) [inline config.yaml:21]: Out of curiosity, I'm wondering the reason for including "none" as one of the options rather than leaving the field empty?

File: lib/charms/smtp_integrator/v0/smtp.py [hunk 27]
```diff
@@ -0,0 +1,292 @@
+#!/usr/bin/env python3
```
Comment (yanksyoon) [inline lib/charms/smtp_integrator/v0/smtp.py:1]: I think the shebang line is not necessary for the library :)

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 19 above)
Comment (merkata) [reply lib/charms/smtp_integrator/v0/smtp.py:73]: The spec says it will provide two interfaces, one legacy and one "standard" or how you call it - I've noticed in the spec the interface is with an underscore - "smtp_legacy", it should be with an underscore here as well to match the spec, right? And as the charm will provide two interfaces, perhaps document that there is a legacy interface and how to use it, but go with the newer one as default? Still for the initial minimal working charm I think this can stay as is, however you prefer.

File: requirements.txt [hunk 28]
```diff
@@ -1 +1,2 @@
-ops >= 2.2.0
+ops==2.7.0
```
Comment (yanksyoon) [inline requirements.txt:1]: May I ask why the specific version pin please? I think there's no need to pin the patch versions :)

File: src/charm.py [hunk 29]
```diff
@@ -33,83 +31,59 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
```
Comment (yanksyoon) [inline src/charm.py:37]: I'm not a too big fan of having exc.msg as the charm status, would "Invalid Charm Status, please see juju debug-logs" be better for longer error messages? 
On the other hand, having explicit messages can be a lot easier to see at first glance though.

**Review by merkata — DISMISSED** [review]
Looks good, I left a comment about smtp-legacy with the suggestion we default to smtp, though I've looked at the code and other comments and I believe that is the intent going forward, so you don't have to answer that :)

**Review by yanksyoon — DISMISSED** [review]
Thank you for the gourmet code, makes my heart feel better!

File: requirements.txt (hunk 28 above)
Comment (arturo-seijas) [reply requirements.txt:1]: Reproducible builds :)

File: src/charm.py (hunk 29 above)
Comment (arturo-seijas) [reply src/charm.py:37]: That was the idea. Under this scenario, we'll be able to see the invalid fields directly in the status

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 27 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py:1]: Indeed. Thx

**Review by amandahla — DISMISSED** [review]
LGTM

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (gregory-schiano) [reply lib/charms/smtp_integrator/v0/smtp.py]: Sorry re-opening, we probably misunderstood. When I said "that's fine" I was approving your approach of instantiating a `SmtpRelationData` object here to trigger validation

**Review by merkata — DISMISSED** [review]

File: lib/charms/smtp_integrator/v0/smtp.py (hunk 12 above)
Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py]: The approach is a bit different. I'm instantiating it when accessing the event data instead of when emitting it for two reasons:
* I don't want to override the ops `emit`method
* This seems more cohesive. The event will be emitted in more places in the future
Let me know your thoughts on this

Comment (arturo-seijas) [reply lib/charms/smtp_integrator/v0/smtp.py]: Changed to instantiate immediately before emitting

**Review by gregory-schiano — APPROVED** [review]
LGTM

**Review by amandahla — APPROVED** [review]
LGTM
